include/libopencm3/pac55xx/ccs.h

```diff
@@ -26,18 +27,95 @@
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  */
-#ifndef INCLUDE_LIBOPENCM3_PAC55XX_CCS_H_
-#define INCLUDE_LIBOPENCM3_PAC55XX_CCS_H_
+#ifndef LIBOPENCM3_PAC55XX_CCS_H_
+#define LIBOPENCM3_PAC55XX_CCS_H_
 
 #include <libopencm3/cm3/common.h>
+#include <libopencm3/cm3/memorymap.h>
+#include <libopencm3/pac55xx/memorymap.h>
+#include <libopencm3/pac55xx/memctl.h>
 
 /**@{*/
 
-/** Clock Control Registers
- * @defgroup clock_config_regs Clock Config Registers.
- * @{*/
+/** @defgroup ccs_frequencies CCS Frequencies
+@{*/
+/** Ring Oscillator Frequency */
+#define CCS_ROSC_FREQ       (16000000U)
+/** Internally generated and trimmed 4MHz clock */
+#define CCS_CLKREF_FREQ     ( 4000000U)
+/** Maximum external clock frequency */
+#define CCS_EXTCLK_MAX_FREQ (20000000U)
+/**@}*/
+
+/** @defgroup ccs_ctl_reg Clock Control Register
+@{*/
 #define CCSCTL MMIO32(SCC_BASE)
+#define CCS_CTL_FRCLKMUXSEL_MASK (0x03)
+#define CCS_CTL_FRCLKMUXSEL(sel) ((sel) & CCS_CTL_FRCLKMUXSEL_MASK)
+#define CCS_CTL_FRCLKMUXSEL_ROSC    (0)
+#define CCS_CTL_FRCLKMUXSEL_CLKREF  (1)
+#define CCS_CTL_FRCLKMUXSEL_EXTCLK  (3)
+#define CCS_CTL_ROSCEN BIT2
+#define CCS_CTL_SCLKMUXSEL BIT4
+#define CCS_CTL_SCLKMUXSEL_FRCLK    (0)
+#define CCS_CTL_SCLKMUXSEL_PLLCLK   (1)
+#define CCS_CTL_CLKFAILEN BIT5
+#define CCS_CTL_CLKFAILMUXSEL BIT6
+#define CCS_CTL_CLKFAILIF BIT7
+#define CCS_CTL_LDOEN BIT8
+#define CCS_CTL_SWRESET BIT11
+#define CCS_CTL_PCLKEN BIT12
+#define CCS_CTL_ACLKEN BIT13
+#define CCS_CTL_ADCCLKEN BIT14
+#define CCS_CTL_STCLKSLPEN BIT15
+#define CCS_CTL_PCLKDIV_MASK (0x07)
+#define CCS_CTL_PCLKDIV_SHIFT (16)
+/* Supported PCLK divisors: 1-8 */
+#define CCS_CTL_PCLKDIV(div) (((div-1) & CCS_CTL_PCLKDIV_MASK) << CCS_CTL_PCLKDIV_SHIFT)
+#define CCS_CTL_ACLKDIV_MASK (0x07)
+#define CCS_CTL_ACLKDIV_SHIFT (20)
+/* Supported ACLK divisors: 1-8 */
+#define CCS_CTL_ACLKDIV(div) (((div-1) & CCS_CTL_ACLKDIV_MASK) << CCS_CTL_ACLKDIV_SHIFT)
+#define CCS_CTL_HCLKDIV_MASK (0x07)
+#define CCS_CTL_HCLKDIV_SHIFT (24)
+/* Supported HCLK divisors: 1-8 */
+#define CCS_CTL_HCLKDIV(div) (((div-1) & CCS_CTL_HCLKDIV_MASK) << CCS_CTL_HCLKDIV_SHIFT)
+#define CCS_CTL_USAMODE BIT28
+#define CCS_CTL_USBMODE BIT29
+#define CCS_CTL_USCMODE BIT30
+#define CCS_CTL_USDMODE BIT31
+/**@}*/
+
+/** @defgroup ccs_pllctl_reg CCS PLL Control Register
+@{*/
 #define CCSPLLCTL MMIO32(SCC_BASE + 0x04)
+/** PLL Enable */
+#define CCS_PLLCTL_PLLEN BIT0
+/** PLL Bypass */
+#define CCS_PLLCTL_PLLBP BIT1
+#define CCS_PLLCTL_PLLOUTDIV_MASK (0x03)
+#define CCS_PLLCTL_PLLOUTDIV_SHIFT (2)
+/** PLL Output Divisor */
+#define CCS_PLLCTL_PLLOUTDIV(div) (((div) & CCS_PLLCTL_PLLOUTDIV_MASK) << CCS_PLLCTL_PLLOUTDIV_SHIFT)
```

Comment:
I was thinking about it, and one way to make this more consistent like the rest of the divisors below (since this is power of 2 instead of incremental), we could do something like this, so the user would only be able to input 1,2,4, or 8, but the register value would still turn out alright, and we'd delete lines 100-104 below.
```
#define CCS_PLLCTL_PLLOUTDIV(div) (((((div + 1) * 2) / 5) & CCS_PLLCTL_PLLOUTDIV_MASK) << CCS_PLLCTL_PLLOUTDIV_SHIFT)
```

Reply:
@karlp Would my suggestion above be consistent with libopencm3 standards?